commands/http/handler.go

```diff
@@ -121,6 +121,16 @@ func (i internalHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	i.ctx.Context = ctx
 	req.SetContext(i.ctx)
 
+	if cn, ok := w.(http.CloseNotifier); ok {
+		go func() {
+			select {
+			case <-cn.CloseNotify():
+				cancel()
+			case <-ctx.Done():
+			}
+		}()
+	}
+
```

Comment:
thats the weirdest thing, the [cast to a channel here](https://github.com/ipfs/go-ipfs/blob/master/core/commands/add.go#L140) is failing because the object isnt actually a channel, its one of the objects that was supposed to be sent on the channel..